food-truck-api/src/main/java/food/truck/api/endpoint/NotificationEndpoint.java

```diff
@@ -0,0 +1,115 @@
+import food.truck.api.truck.TruckService;
+import food.truck.api.user.User;
+import lombok.NonNull;
+import lombok.Value;
+import lombok.extern.log4j.Log4j2;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.access.annotation.Secured;
+import org.springframework.security.core.annotation.AuthenticationPrincipal;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+@Log4j2
+@RestController
+public class NotificationEndpoint {
+    @Autowired
+    private NotificationService notificationService;
+
+    @Autowired
+    private SubscriptionService subscriptionService;
+
+    @Autowired
+    TruckService truckService;
+
+    @Value
+    public static class SendNotificationParams {
+        long truckId;
+        @NonNull
+        String message;
+        // TODO: different types of notifications
+    }
+
+    @Secured({"ROLE_USER"})
+    @PostMapping("/truck/notification")
+    public boolean sendNotificationForTruck(
+        @AuthenticationPrincipal User u,
+        @RequestBody SendNotificationParams notificationParams
+    ) {
+        var t = truckService.findTruckById(notificationParams.truckId);
+        if (t.isEmpty()) {
+            return false;
+        }
+        var truck = t.get();
+        if (!truck.getUserId().equals(u.getId()) && u.getIsOwner()) {
+            return false;
+        }
+        var subscriptions = subscriptionService.findSubsByTruck(truck);
+        notificationService.saveNotification(subscriptions, notificationParams.message);
+        return true;
+    }
+
+    @Secured({"ROLE_USER"})
+    @GetMapping("/truck/{truckId}/notifications")
+    public List<NotificationView> sendNotificationForTruck(
+        @AuthenticationPrincipal User u,
+        @PathVariable long truckId
+    ) {
+        var t = truckService.findTruckById(truckId);
+        if (t.isEmpty()) {
+            return List.of();
+        }
+        var truck = t.get();
+        if (!truck.getUserId().equals(u.getId()) && u.getIsOwner()) {
+            return List.of();
+        }
+        return notificationService.findNotificationsByTruck(truck);
+    }
+
+    @Secured({"ROLE_USER"})
+    @GetMapping("/notifications")
+    public List<NotificationView> getNotificationsForUser(@AuthenticationPrincipal User user) {
+        return notificationService.findNotificationsByUser(user);
+    }
+
+    @Value
+    public static class UpdateNotificationStatusParams {
+        long notificationId;
+        boolean isRead;
+    }
+
+    @Secured({"ROLE_USER"})
+    @PutMapping("/notification/read")
+    public void setNotificationReadStatus(
+        @AuthenticationPrincipal User user,
+        @RequestBody UpdateNotificationStatusParams updateStatus
+    ) {
+        var notification = notificationService.findById(updateStatus.notificationId);
+        notification.ifPresent(n -> {
+            if (n.getSubscription().getUser().getId().equals(user.getId())) {
```

Comment:
Not 100% sure this work should be done in the endpoint, looks like work for the service. But I'll leave that up to you.

Reply:
I had just done it this way since I was thinking of having the endpoint also serve as a "filter" for who has authorization for calling service methods on certain notifications